build-support/migration-support/migrate_to_toml_config.py

```diff
@@ -0,0 +1,127 @@
+    joined_new_content = '\n'.join(new_content) + "\n"
+    if args.preview:
+      print(f"Would create {new_path} with the following content:\n\n{joined_new_content}")
+    else:
+      logging.info(
+        f"Created {new_path}. There are likely some remaining issues that need manual "
+        "attention. Please copy the file into https://www.toml-lint.com or open with your editor "
+        "to fix any remaining issues."
+      )
+      new_path.write_text(joined_new_content)
+
+
+def create_parser() -> argparse.ArgumentParser:
+  parser = argparse.ArgumentParser(
+    description='Convert INI config files to TOML config files.',
+  )
+  parser.add_argument(
+    'files', type=Path, nargs='*',
+    help="Config files to convert.",
+  )
+  parser.add_argument(
+    '-p', '--preview', action='store_true',
+    help="Output to stdout rather than creating the new TOML config file(s).",
+  )
+  return parser
+
+
+def update_primitive_value(original: str) -> str:
+  if original in ["true", "True"]:
+    return "true"
+  if original in ["false", "False"]:
+    return "false"
+
+  try:
+    return str(int(original))
+  except ValueError:
+    pass
+
+  try:
+    return str(float(original))
+  except ValueError:
+    pass
+
+  return f'"{original}"'
+
+
+def generate_new_config(config: Path) -> List[str]:
+  original_text = config.read_text()
+  original_text_lines = original_text.splitlines()
+  updated_text_lines = original_text_lines.copy()
+
+  for i, line in enumerate(original_text_lines):
+    option_regex = r"(?P<option>[a-zA-Z0-9_]+)"
+    before_value_regex = rf"\s*{option_regex}\s*[:=]\s*"
+    valid_value_characters = r"a-zA-Z0-9_.@!%\=\>\<\-\(\)\/"
+    value_regex = rf"(?P<value>[{valid_value_characters}]+)"
+    parsed_line = re.match(
+      rf"{before_value_regex}{value_regex}\s*$", line,
+    )
+
+    if parsed_line:
+      option, value = parsed_line.groups()
+      updated_text_lines[i] = f"{option} = {update_primitive_value(value)}"
+      continue
+
+    # Check if it's a one-line list value
+    list_value_regex = rf"(?P<list>[\+\-]?\[[{valid_value_characters},\s\'\"]*\])"
+    parsed_list_line = re.match(rf"{before_value_regex}{list_value_regex}\s*$", line)
+    if parsed_list_line:
+      option, value = parsed_list_line.groups()
+      if value.startswith("+"):
+        updated_line = f"{option}.append = {value[1:]}"
+      elif value.startswith("-"):
+        updated_line = f"{option}.filter = {value[1:]}"
+      else:
+        updated_line = f"{option} = {value}"
+      updated_text_lines[i] = updated_line
+      continue
+
+    # Check if it's a one-line dict value
+    dict_value_regex = rf"(?P<dict>{{[{valid_value_characters},:\s\'\"]*}})"
+    parsed_dict_line = re.match(rf"{before_value_regex}{dict_value_regex}\s*$", line)
+    if parsed_dict_line:
+      option, value = parsed_dict_line.groups()
+      updated_text_lines[i] = f'{option} = """{value}"""'
+      continue
+
```

Comment:
maybe I'll go on a limb here and suggest some plaintext code:
This is supposed to go after the last if and before the return
```suggestion
    print(f'We couldn't automatically convert line {i}: {line}. Please refer to https://github.com/pantsbuild/pants/pull/9054 for a description of what might be the case.')
```

Reply:
Thank you for the code snippet!

The one challenge with this snippet is that it will trigger on every non-option line, such as every single one of these lines, except for the very last one:
```ini
# Comment not attached to any option
[cache.java]
args: [
     'foo',
     'bar'
  ]


next_option: foo
```

I think that's more output than we would want. We could, of course, teach the script how to handle those different things like what a section header looks like, but then we're getting to the complexity I wanted to avoid. 

--

Favor: can you please pull down this PR and run `build-support/migration-support/migrate_to_toml_config.py pants.ini pants.travis-ci.ini pants.remote.ini`, then open up the new TOML files in your editor? (Or, do this on Twitter's config files). 

I suspect you'll find that there aren't many remaining issues and that those errors are fairly intuitive to fix, thanks to the editor warnings. This is why I'm pushing back on going from 80-100%; after trying out the script, it doesn't seem worth it to make it even better.

@benjyw's point is that we don't even need to provide a script in the first place - it's not very difficult to convert from INI to TOML, only a little tedious. We're providing this script as a courtesy to users, and I don't think we need to have it be perfect for them to be happy.

Reply:
I still think I'd prefer to be overly verbose, but I agree that nobody is going to die for figuring out 20 manual changes. Particularly, because "users" here most likely means the same people that are seeing these changes happen, not actual end-users.

So not going to block on that, thanks for putting the effort to consider this point :)